components/server/src/auth/auth-provider-service.ts

```diff
@@ -56,73 +60,268 @@ export class AuthProviderService {
+            return undefined;
+        }
+
+        if (result.organizationId) {
+            await this.auth.checkPermissionOnOrganization(userId, "write_git_provider", result.organizationId);
         } else {
-            const existing = await this.authProviderDB.findByHost(entry.host);
-            if (existing) {
-                throw new Error("Provider for this host already exists.");
-            }
-            authProvider = this.initializeNewProvider(entry);
+            await this.auth.checkPermissionOnUser(userId, "write_info", userId);
         }
-        return await this.authProviderDB.storeAuthProvider(authProvider as AuthProviderEntry, true);
+
+        return result;
     }
 
-    async createOrgAuthProvider(entry: AuthProviderEntry.NewOrgEntry): Promise<AuthProviderEntry> {
-        const orgProviders = await this.authProviderDB.findByOrgId(entry.organizationId);
-        const existing = orgProviders.find((p) => p.host === entry.host);
+    async createAuthProviderOfUser(userId: string, entry: AuthProviderEntry.NewEntry): Promise<AuthProviderEntry> {
+        await this.auth.checkPermissionOnUser(userId, "write_info", userId);
+
+        const host = entry.host && entry.host.toLowerCase();
+
+        // reachability test
+        if (!(await this.isHostReachable(host))) {
+            log.info(`Host could not be reached.`, { entry });
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, `Host could not be reached.`);
+        }
+
+        // checking for already existing runtime providers
+        const isBuiltInProvider = this.isBuiltInProvider(host);
+        if (isBuiltInProvider) {
+            log.info(`Attempt to override an existing provider.`, { entry });
+            throw new ApplicationError(ErrorCodes.CONFLICT, `Attempt to override an existing provider.`);
+        }
+        const existing = await this.authProviderDB.findByHost(entry.host);
         if (existing) {
-            throw new Error("Provider for this host already exists.");
+            log.info(`Provider for this host already exists.`, { entry });
+            throw new ApplicationError(ErrorCodes.CONFLICT, `Provider for this host already exists.`);
         }
 
         const authProvider = this.initializeNewProvider(entry);
+        return await this.authProviderDB.storeAuthProvider(authProvider as AuthProviderEntry, true);
+    }
+
+    private isBuiltInProvider(host: string) {
+        return this.config.authProviderConfigs.some((config) => config.host.toLowerCase() === host.toLocaleLowerCase());
+    }
+
+    async updateAuthProviderOfUser(userId: string, entry: AuthProviderEntry.UpdateEntry): Promise<AuthProviderEntry> {
+        await this.auth.checkPermissionOnUser(userId, "write_info", userId);
+
+        const { id, ownerId } = entry;
+        const existing = (await this.authProviderDB.findByUserId(ownerId)).find((p) => p.id === id);
+        if (!existing) {
+            throw new Error("Provider does not exist.");
```

Comment:
we should use ApplicationErrror, all standard errors are considered internal errors

Reply:
Yeah, you are right. 

BTW, that mentioned default behavior was changed recently, and broke in several places. We should avoid introducing default behavior without aligning _all_ services.

I'll fix the returns of this service.